darshan-util/pydarshan/darshan/experimental/plots/plot_dxt_heatmap.py

```diff
@@ -0,0 +1,440 @@
+
+
+def plot_dxt_heatmap(
+    report,
+    mods=None,
+    ops=None,
+    xbins=200,
+):
+    if mods is None:
+        mods = ["DXT_POSIX"]
+    elif mods[0] != "DXT_POSIX":
+        # TODO: for the moment reject any cases that don't input "DXT_POSIX"
+        # until we can properly aggregate the data
+        raise NotImplementedError(f"DXT heatmap not implemented for module {mods[0]}")
+    if ops is None:
+        ops = ["read", "write"]
+
+    # generate the flattened data dictionary from the report data
+    flat_data_dict = retrieve_flat_data(report=report, mods=mods, ops=ops)
+    # aggregate the data according to the selected modules and operations
+    start_time_data, end_time_data, ranks_data, bytes_data = aggregate_data(
+        data=flat_data_dict, mods=mods, ops=ops
+    )
+
+    # get the heatmap data array
+    Hmap_data = get_heatmap_data(
+        start_time_data=start_time_data,
+        end_time_data=end_time_data,
+        ranks_data=ranks_data,
+        data_arr=bytes_data,
+        n_xbins=xbins,
+    )
+
+    # get the unique ranks
+    unique_ranks = np.unique(ranks_data)
+    ybins = unique_ranks.size
+
+    # create a 2D histogram from the time and rank data, with the
+    # selected data as the weight
+    bins = [xbins, ybins]
+    jgrid = sns.jointplot(kind="hist", bins=bins, space=0.05)
+    jgrid.ax_marg_x.cla()
+    jgrid.ax_marg_y.cla()
+
+    colorbar_label = f"Data (B): {', '.join(ops)}"
+    colorbar_kws = {"label": colorbar_label}
+    hmap = sns.heatmap(
+        Hmap_data,
+        ax=jgrid.ax_joint,
+        # choose a color map that is not white at any value
+        cmap="YlOrRd",
+        norm=LogNorm(),
+        cbar_kws=colorbar_kws,
+    )
+
+    for _, spine in hmap.spines.items():
+        spine.set_visible(True)
+
+    if unique_ranks.size > 1:
+        jgrid.ax_marg_y.barh(
+            y=unique_ranks,
+            width=Hmap_data.sum(axis=1),
+            align="edge",
+            height=-0.99,
+            facecolor="black",
+            lw=0.5,
+        )
+    else:
+        jgrid.ax_marg_y.axis("off")
+
+    jgrid.ax_marg_x.bar(
+        x=np.arange(xbins),
+        height=Hmap_data.sum(axis=0),
+        facecolor="black",
+        align="edge",
+    )
+
+    # only create 4 x-axis tick marks, for figure clarity
+    n_xlabels = 4
+    # create x ticks that scale with the number of xbins in the 2D histogram
+    xticks = np.linspace(0, xbins, n_xlabels)
+    jgrid.ax_joint.set_xticks(xticks)
+    # for the x tick labels, start at 0 and end with the max time (converted to
+    # an integer)
+    max_time = np.max(end_time_data)
+    if max_time < 1:
+        x_ticklabels = np.around(np.linspace(0, max_time, n_xlabels), decimals=2)
+    elif (max_time > 1) & (max_time < 10):
+        x_ticklabels = np.around(np.linspace(0, max_time, n_xlabels), decimals=1)
+    else:
+        x_ticklabels = np.asarray(
+            np.linspace(0, np.ceil(max_time), n_xlabels), dtype=int
```

Comment:
`linspace` accepts a `dtype` argument and returns an array, so why wrap it in an `asarray` call that coerces dtype and checks if it is an array?